### PR TITLE
Fix newline error in solidity::util::readUntilEnd

### LINKED ISSUE #11553: [CLI] Different source hashes when source is passed via standard input and via command line parameter
## Description
The content of the file passed to standard input does not pass through unmodified and its content hash stored in metadata (note: I do not mean the metadata hash itself) is different than when the file name is given as an argument.

The problem is caused by our function for reading standard input, which replaces all newlines with `\n` and also adds a newline after the last line, even if originally there wasn't one.

In practice the consequences of this are probably minor because even if the source hashes end up being the same, metadata won't be - due to different file names (unless the file is called `<stdin>` which is unlikely). Still, I think that this behavior is unexpected by users and should be changed. It looks intentional though and might be a workaround for something so we should discuss it first.

## Environment

- Compiler version: 0.8.5
- Operating system: Arch Linux

## Steps to Reproduce
```bash
echo "contract C {}" > '<stdin>'
solc '<stdin>' --metadata
```
`sources` key from metadata output:
```json
    "sources": {
        "<stdin>": {
            "keccak256": "0xbe86d3681a198587296ad6d4a834606197e1a8f8944922c501631b04e21eeba2",
            "urls": [
                "bzz-raw://7ef32f5b5f3b1ebe447f46969c01ce0e9190d0b90241cbc228255866326be8be",
                "dweb:/ipfs/QmZRepHCmeSSTfxZvKU5KS7pebBYT4khWhTnqh6GzBFU3M"
            ]
        }
    },
```
```bash
cat contract.sol | solc - --metadata
```
`sources` key from metadata output:
```json
    "sources": {
        "<stdin>": {
            "keccak256": "0xec5627577fc5b8010229a1036827ef0739fb21bb05b7bf259947acdded145877",
            "urls": [
                "bzz-raw://2363616d36c4824b30b1698ffa5bc3443635a4d68eb4a93c7126b30b7716f0a1",
                "dweb:/ipfs/QmYENep2eioJEAMKZWozLW3fb2BHjNxCXSszbDzz2kfWJu"
            ]
        }
    },
```

### PR DESCRIPTION
**Pull request by TerranCivilian**
Fixes #11553

### CODE REVIEW

**Comment by TerranCivilian** [comment]
Oh... everything passed in soltest on my original branch. I naively did not run soltest again after rebasing; didn't think I'd need to. But I'm getting soltest errors after the rebase onto develop.

edit: Oh, soltest is failing on my local develop branch again; I dunno what keeps happening. I am thoroughly confused at this point.

**Comment by cameel** [comment]
It's not `soltest`. It's the command-line tests. Try running them locally with `test/cmdlineTests.sh`.

I see that the failure happens while compiling one of the snippets from documentation (specifically one from the page on [Inline Assembly](https://docs.soliditylang.org/en/latest/assembly.html)). Until recently we have been testing only Solidity snippets but just before the weekend a PR that adds testing for Yul snippets too has been merged (#11616). It did pass all tests but there's always a possibility that some other PR went in the meantime and tests do not pass after merge. Please check if it fails also on `develop` or only in your branch.

To debug it, take a look at the command the script runs:
```bash
"/root/project/build/solc/solc" --strict-assembly /tmp/tmp.dc0WgaSphP/test_4111f9d52bcf44f413fb3fcc0ff8e02cb0c1715326b663fb25548cd05709b13a_assembly_rst.yul
```

You should be able to run that manually and check what's happening. `cmdlineTests.sh` should leave the files in `/tmp` it if fails. But if it does not, you can also get the set of input files it uses by running `scripts/isolate_tests.sh`.

File: libsolutil/CommonIO.h [hunk 1]
```diff
@@ -39,7 +39,7 @@ namespace solidity::util
 std::string readFileAsString(std::string const& _file);
 
 /// Retrieves and returns the whole content of the specified input stream (until EOF).
-std::string readUntilEnd(std::istream& _stdin);
+std::string readUntilEnd(std::istream const& _stdin);
```
Comment (cameel) [inline libsolutil/CommonIO.h]: Interesting that it actually compiles with `const` here. Probably because by accessing `rdbuf()` you're not marking the text you read a actually read. I wonder if it might be better to flush the stream so that other parts of the compiler cannot accidentally process the same input again.

File: test/libsolutil/CommonIO.cpp [hunk 2]
```diff
@@ -64,6 +64,21 @@ BOOST_AUTO_TEST_CASE(readFileAsString_symlink)
...
+BOOST_AUTO_TEST_CASE(readUntilEnd_with_ending_newline)
+{
+	BOOST_TEST(readUntilEnd(istringstream("ABC\ndef\n")) == "ABC\ndef\n");
+}
+
+BOOST_AUTO_TEST_CASE(readUntilEnd_empty)
+{
+	BOOST_TEST(readUntilEnd(istringstream("")) == "");
+}
+
```
Comment (cameel) [inline test/libsolutil/CommonIO.cpp]: Can you also add a test case with Windows line endings? That's the case that would trip up your previous implementation so better have it covered.

**Review by cameel — COMMENTED** [review]
Looks fine to me overall. Not sure about the `const` stream but I don't think it breaks anything right now (we're only reading from the stream once) so maybe it does not matter all that much.

File: libsolutil/CommonIO.h (hunk 1 above)
Comment (christianparpart) [reply libsolutil/CommonIO.h]: It should not be `const` anyhow. Otherwise the caller might think it's actually not consuming from `_stdin`?

Comment (cameel) [reply libsolutil/CommonIO.h]: But it's not consuming stdin content :) I think you can still read it after calling this function because `rdbuf()` just gives you the internal pointer to the content of the stream.

The question is just whether it's a good thing or not. I think that currently the name implies that the content would be consumed so I think I'd be in favor of making it non-`const` and making it clear the buffer.

**Comment by Marenz** [comment]
The command line test failure is not due to your PR. Try rebasing on #11644 (or wait till it is merged into develop), that should fix it.

File: libsolutil/CommonIO.h (hunk 1 above)
Comment (christianparpart) [reply libsolutil/CommonIO.h]: `peekUntilEnd()`?

Comment (cameel) [reply libsolutil/CommonIO.h]: I think that would be `peekUntilEnd()`. But other than that it sounds good to me. I originally wanted to call it `captureInputStream()` when I refactored it recently. I like "peek" though because it would immediately tell me that the function does not consume stuff from the stream.

Comment (christianparpart) [reply libsolutil/CommonIO.h]: @cameel also, "peek" is a common term in text processing and known not to consume (as opposed to "get", "read", "consume", etc.)

Comment (cameel) [reply libsolutil/CommonIO.h]: Yeah. That's what I meant by saying that it tells me immediately what actually happens :)

Comment (christianparpart) [reply libsolutil/CommonIO.h]: @cameel Another concern I have: if we keep `readUntilEnd(istream&)`, then it should definitely consume the input (we can still add a `peekUntilEnd()`.

**Comment by cameel** [comment]
@TerranCivilian Are you still working on the PR? The current solution is good in my opinion so the only thing this needs before it can get merged is a rebase on `develop`.

**Comment by alice-dfi** [comment]
It better not be anything illegal or you will never have a chance to do this or anything with code in the future .

**Comment by TerranCivilian** [comment]
> @TerranCivilian Are you still working on the PR? The current solution is good in my opinion so the only thing this needs before it can get merged is a rebase on `develop`.

Hey, sorry, I've recently been much more busy outside of work. I can try to finish now. I never really figured out how my branch got messed up or why the CircleCI tests were failing. I think I will just re-clone.

**Comment by cameel** [comment]
@TerranCivilian It turned out that the breakage was caused by an unrelated PR merged into `develop`. As I said above, simply rebasing your PR on `develop` should be enough to make tests pass get make your PR mergeable.

**Comment by TerranCivilian** [comment]
Hm, I think I messed up again; here's exactly what I just did:

1. Went to my GitHub solidity repo at https://github.com/TerranCivilian/solidity
2. Fetched + merged upstream
3. Went to my local repo with develop checkout out and did:
  a. `git fetch`
  b. `git pull`
  c. `git checkout src-hash-CommonIO-fix`
  d. `git rebase develop`
  e. `git push -f`

Did I do it wrong? Or is develop messed up again

**Comment by cameel** [comment]
No, it's all fine. It actually proves that your PR does its job :) Tests are failing because my PRs went into `develop` in the meantime and they added some tests that include the expected content of stdin among other things. Now that your PR is on top of that, you need to update expectations in these tests - they should no longer expect a newline on stdin - they can assume it's empty when nothing was passed on it.

So basically just go over `t_ubu_soltest`, look at where the problem is in each failing case (there's a file name and line number) and update the test.

**Comment by cameel** [comment]
> a. `git fetch`
> b. `git pull`

You do not need both. `pull` is just `fetch` + `merge`. I personally never use `pull`. `fetch` is convenient because it's safe to do at any time (it does not modify my checked out branch) and then I can just merge if I see that the branch can be safely fast forwarded.

> d. `git rebase develop`

Use `git rebase develop --interactive`. It will let you review commits before rebase and you can abort if something's not right.

This way you can also drop commits that are not yours - there could be some if your PR is based on some other branch and this branch is force pushed with modified commits. git would then think that those old versions of commits in your branch are actually your own commits and include them in the rebase. Without `--interactive` (or even with `--interactive if you do not drop these commits) you'll just get weird conflicts in that case.

> e. `git push -f`

I'd recommend to always use `git push --force-with-lease` instead of `--force`. It ensures that if someone pushed to your branch in the meantime, git will tell you instead of silently overwriting his changes.

**Comment by cameel** [comment]
Just noticed that your PR is missing a changelog entry. You should mention your fix in [Changelog.md](https://github.com/ethereum/solidity/blob/develop/Changelog.md).

**Comment by cameel** [comment]
I have rebased this on `develop` and fixed test expectations. If tests pass, I think we can merge this.

File: test/libsolutil/CommonIO.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolutil/CommonIO.cpp]: We should test this. Maybe it's also different depending on the platform?

If newlines were transformed on windows, this should get a changelog entry.

File: libsolutil/CommonIO.h (hunk 1 above)
Comment (chriseth) [reply libsolutil/CommonIO.h]: Is there any situation where we would want this function to consume the input?

Comment (cameel) [reply libsolutil/CommonIO.h]: Not really. For me it's just that from the name I would expect that it does consume it.

Changed the name to `peekUntilEnd()`.

File: test/libsolutil/CommonIO.cpp (hunk 2 above)
Comment (cameel) [reply test/libsolutil/CommonIO.cpp]: I checked and looks like it was not converting newlines after all. It's because the line ending that `getline()` strips is always `\n` and not platform-dependent as I expected.

So the only thing the previous implementation was doing wrong was always adding an extra `\n` at the end of the file. Added a changelog entry saying this.

**Comment by chriseth** [comment]
Conflict in changelog.

**Comment by cameel** [comment]
Fixed.

File: test/libsolutil/CommonIO.cpp (hunk 2 above)
Comment (cameel) [reply test/libsolutil/CommonIO.cpp]: Looks like outputting the `rdbuf()` does advance the stream (the function is `const` but the pointer it returns is not) so I changed the parameter back to a non-`const` reference and renamed the function back to `readUntilEnd()`.

File: libsolutil/CommonIO.h (hunk 1 above)
Comment (cameel) [reply libsolutil/CommonIO.h]: Looks like outputting the `rdbuf()` does advance the stream (the function is `const` but the pointer it returns is not) so I changed the parameter back to a non-`const` reference and renamed the function back to `readUntilEnd()`.

**Review by chriseth — APPROVED** [review]
